File: 목요일스터디/BOJ20327/김재환20327.java
import java.io.*;
import java.util.*;

public class Main {

    static int N, R, size;
    static int[][] A;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        R = Integer.parseInt(st.nextToken());

        size = 1 << N;
        A = new int[size][size];

        for (int i = 0; i < size; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < size; j++) {
                A[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        while (R-- > 0) {
            st = new StringTokenizer(br.readLine());
            int k = Integer.parseInt(st.nextToken());
            int l = Integer.parseInt(st.nextToken());

            applyOperation(k, l);
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < size; j++) {
                sb.append(A[i][j]).append(' ');
            }
            sb.append('\n');
        }
        System.out.print(sb);
    }

    static void applyOperation(int k, int l) {
        int s = 1 << l;        // block size
        int nb = size / s;    // number of blocks per row
        int[][] B = new int[size][size];

        for (int i = 0; i < size; i++) {
            int bi = i / s;
            int ii = i % s;

            for (int j = 0; j < size; j++) {
                int bj = j / s;
                int jj = j % s;

                int nbi = bi, nbj = bj;
                int nii = ii, njj = jj;

                switch (k) {
                    case 1: // block 내부 상하 반전
                        nii = s - 1 - ii;
                        break;
                    case 2: // block 내부 좌우 반전
                        njj = s - 1 - jj;
                        break;
                    case 3: // block 내부 시계 회전
                        nii = jj;
                        njj = s - 1 - ii;
                        break;
                    case 4: // block 내부 반시계 회전
                        nii = s - 1 - jj;
                        njj = ii;
                        break;
                    case 5: // block 상하 반전
                        nbi = nb - 1 - bi;
                        break;
                    case 6: // block 좌우 반전
                        nbj = nb - 1 - bj;
                        break;
                    case 7: // block 시계 회전
                        nbi = bj;
                        nbj = nb - 1 - bi;
                        break;
                    case 8: // block 반시계 회전
                        nbi = nb - 1 - bj;
                        nbj = bi;
                        break;
                }

                int ni = nbi * s + nii;
                int nj = nbj * s + njj;
                B[ni][nj] = A[i][j];
            }
        }

        A = B;
    }
}

File: 목요일스터디/BOJ20327/김재환20327.py
import sys
input = sys.stdin.readline

def solve():
    N, R = map(int, input().split())
    size = 1 << N  # 2^N

    A = [list(map(int, input().split())) for _ in range(size)]

    for _ in range(R):
        k, l = map(int, input().split())
        s = 1 << l          # block size = 2^l
        nb = size // s      # number of blocks per row/col

        B = [[0] * size for _ in range(size)]

        for i in range(size):
            bi, ii = divmod(i, s)   # block row, in-block row
            for j in range(size):
                bj, jj = divmod(j, s)  # block col, in-block col

                # default: block stays, inside stays
                nbi, nbj = bi, bj
                nii, njj = ii, jj

                if k == 1:      # flip inside block vertically
                    nii = s - 1 - ii
                elif k == 2:    # flip inside block horizontally
                    njj = s - 1 - jj
                elif k == 3:    # rotate inside block right
                    nii, njj = jj, s - 1 - ii
                elif k == 4:    # rotate inside block left
                    nii, njj = s - 1 - jj, ii
                elif k == 5:    # flip blocks vertically
                    nbi = nb - 1 - bi
                elif k == 6:    # flip blocks horizontally
                    nbj = nb - 1 - bj
                elif k == 7:    # rotate blocks right
                    nbi, nbj = bj, nb - 1 - bi
                elif k == 8:    # rotate blocks left
                    nbi, nbj = nb - 1 - bj, bi

                ni = nbi * s + nii
                nj = nbj * s + njj
                B[ni][nj] = A[i][j]

        A = B

    out = []
    for row in A:
        out.append(" ".join(map(str, row)))
    print("\n".join(out))

if __name__ == "__main__":
    solve()
